### PR TITLE
SONAR-7269 Split installing plugins into installing and updating in api/plugins/pending

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-7269 Split installing plugins into installing and updating in api/plugins/pending

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java [hunk 1]
```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
...
+      .filter(new MatchPluginKeys(installedPluginKeys))
+      .toList();
+    pluginWSCommons.writePluginInfoList(json, updatedPlugins, compatiblePluginsByKey, ARRAY_UPDATING);
+  }
+
+  private enum PluginInfoToKey implements Function<PluginInfo, String> {
+    INSTANCE;
+
+    @Override
+    public String apply(@Nonnull PluginInfo input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java [hunk 2]
```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
...
-    pluginWSCommons.writePluginInfoList(json, plugins, compatiblePluginsByKey, ARRAY_REMOVING);
+  private static class MatchPluginKeys implements Predicate<PluginInfo> {
+    private final List<String> pluginKeys;
+
+    private MatchPluginKeys(List<String> pluginKeys) {
+      this.pluginKeys = pluginKeys;
+    }
+
+    @Override
+    public boolean apply(@Nonnull PluginInfo input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java [hunk 3]
```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
...
+    public String apply(@Nonnull PluginInfo input) {
+      return input.getKey();
+    }
   }
 
-  private void writeRemoving(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
-    Collection<PluginInfo> plugins = installer.getUninstalledPlugins();
-    pluginWSCommons.writePluginInfoList(json, plugins, compatiblePluginsByKey, ARRAY_REMOVING);
+  private static class MatchPluginKeys implements Predicate<PluginInfo> {
+    private final List<String> pluginKeys;
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: a Set should be used to repeatedly call the `contains` method

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java [hunk 4]
```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
...
-    pluginWSCommons.writePluginInfoList(json, plugins, compatiblePluginsByKey, ARRAY_INSTALLING);
+  private void writePlugins(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
+    Collection<PluginInfo> uninstalledPlugins = installer.getUninstalledPlugins();
+    pluginWSCommons.writePluginInfoList(json, uninstalledPlugins, compatiblePluginsByKey, ARRAY_REMOVING);
+
+    Collection<PluginInfo> downloadedPlugins = pluginDownloader.getDownloadedPlugins();
+    Collection<PluginInfo> installedPlugins = installer.getPluginInfos();
+    List<String> installedPluginKeys = from(installedPlugins).transform(PluginInfoToKey.INSTANCE).toList();
+
+    Collection<PluginInfo> newPlugins = from(downloadedPlugins)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: I think I prefer one loop that feeds 2 lists `newPlugins` and `updatedPlugins`. You could call the `MatchPluginKeys` predicate explicitly of just create a test method.

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: or if you want to use functional constructs, I would use a Function that transform one List in 2 Lists (you would need to create an internal data class)

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: ok

**Comment by teryk** [comment]
apply feedback and LGTM

**Pull request merged**